File: Fibonacci.java
public class Fibonacci {
    /**
     * Calcula o número de Fibonacci de forma recursiva
     * @param n Número maior ou igual a 0
     * @return Valor correspondente na sequência de Fibonacci
     * @throws IllegalArgumentException se o número for negativo
     */
    public static int fibonacciRecursivo(int n) {
        if (n < 0) {
            throw new IllegalArgumentException("O número deve ser maior ou igual a 0");
        }
        
        if (n <= 1) {
            return n;
        }
        
        return fibonacciRecursivo(n - 1) + fibonacciRecursivo(n - 2);
    }

    /**
     * Calcula o número de Fibonacci de forma iterativa
     * @param n Número maior ou igual a 0
     * @return Valor correspondente na sequência de Fibonacci
     * @throws IllegalArgumentException se o número for negativo
     */
    public static int fibonacciLinear(int n) {
        if (n < 0) {
            throw new IllegalArgumentException("O número deve ser maior ou igual a 0");
        }
        
        if (n <= 1) {
            return n;
        }
        
        int a = 0;
        int b = 1;
        for (int i = 2; i <= n; i++) {
            int temp = a + b;
            a = b;
            b = temp;
        }
        return b;
    }

    public static void main(String[] args) {
        // Testando a função recursiva
        System.out.println("Usando função recursiva:");
        for (int i = 0; i < 7; i++) {
            System.out.printf("fibonacciRecursivo(%d) = %d%n", i, fibonacciRecursivo(i));
        }
        
        System.out.println("\nUsando função linear:");
        for (int i = 0; i < 7; i++) {
            System.out.printf("fibonacciLinear(%d) = %d%n", i, fibonacciLinear(i));
        }
    }
} 

File: Fibonacci.py
def fibonacci_recursivo(n):
    """Calcula o número de Fibonacci de forma recursiva"""
    if not isinstance(n, int) or n < 0:
        raise ValueError("O número deve ser um inteiro maior ou igual a 0")
    
    if n <= 1:
        return n
    
    return fibonacci_recursivo(n - 1) + fibonacci_recursivo(n - 2)

def fibonacci_linear(n):
    """Calcula o número de Fibonacci de forma iterativa"""
    if not isinstance(n, int) or n < 0:
        raise ValueError("O número deve ser um inteiro maior ou igual a 0")
    
    if n <= 1:
        return n
    
    a, b = 0, 1
    for _ in range(2, n + 1):
        a, b = b, a + b
    return b

# Testes
if __name__ == "__main__":
    print("Usando função recursiva:")
    for i in range(7):
        print(f"fibonacci_recursivo({i}) = {fibonacci_recursivo(i)}")
    
    print("\nUsando função linear:")
    for i in range(7):
        print(f"fibonacci_linear({i}) = {fibonacci_linear(i)}")
